[FIX] resource_calendar: ignore past or unconfirmed bookings when updating calendars

yajo opened the pull request:
Without this patch, users couldn't change a calendar schedule if there were past or unconfirmed bookings that wouldn't fit in it.

Excluding those bookings from the check fixes the situation.

We also check that, to confirm a booking, it must fit in the calendar (because now it can happen that, in the time that has passed since the booking was scheduled until it is confirmed, the calendar changes).


@Tecnativa TT29509

pedrobaeza commented:
I don't get the use case if this leads to something forbidden...

yajo commented:
It was taking into consideration past bookings too. Do you think it's better to narrow the search?

pedrobaeza commented:
Although past events, if they were booked, why now ignore them? Shouldn't they be removed first?

yajo commented:
No, there's no reason to remove past bookings.

Just we shouldn't forbid altering the calendar just because some booking in the past didn't fit in it. The calendar is only used to allow bookings, but you can't book something for the past, only for the future.

pedrobaeza commented:
OK, then forbid only for booking on the future, yes

yajo commented:
OK, I changed the implementation and instead ignored past and unconfirmed events. I hide previous conversation and changed description and title, as this is completely different. Please review again. Thanks everyone.

pedrobaeza commented:
I'm still not sure about the unconfirmed part, as they can conflict when confirming them.

yajo commented:
Yes, that's the goal.

Imagine your customer prereserved some slot, but then you change the calendar, and later the customer tries to confirm. IMHO:

- the system shouldn't forbid you altering the calendar just because your customer didn't finish the process.
- the customer should get an error if confirming late, and thus he should reschedule.

pedrobaeza commented:
But such error confirming the booking exists?

yajo commented:
Yes, but let me add that assertion to the test.

Commit:
[FIX] resource_calendar: ignore past or unconfirmed bookings when updating calendars

Without this patch, users couldn't change a calendar schedule if there were past or unconfirmed bookings that wouldn't fit in it.

Excluding those bookings from the check fixes the situation.

We also check that, to confirm a booking, it must fit in the calendar (because now it can happen that, in the time that has passed since the booking was scheduled until it is confirmed, the calendar changes).

@Tecnativa TT29509

yajo commented:
Oops, I was wrong. Indeed the test showed that it could happen. I added the fix and the test.

pedrobaeza commented:
/ocabot merge patch
